spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/Hack.java

```diff
@@ -0,0 +1,450 @@
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.reactor;
+
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.reactivestreams.Processor;
+import org.reactivestreams.Publisher;
+import reactor.core.CoreSubscriber;
+import reactor.core.Disposable;
+import reactor.core.Fuseable;
+import reactor.core.Scannable;
+import reactor.core.publisher.ConnectableFlux;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxOperator;
+import reactor.core.publisher.GroupedFlux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoOperator;
+import reactor.core.publisher.Operators;
+import reactor.core.publisher.ParallelFlux;
+import reactor.util.annotation.Nullable;
+
+/**
+ * Helper class for reactor ON_EACH instrumentation. Reduces number of sleuth operators
+ * time in the reactive chain during Assembly by using {@code Scannable.Attr.RUN_STYLE}.
+ * Below is the example of what it does <pre>{@code
+ *	Mono.just(0) // (-)
+ * 	.map(it -> 1) // (-)
+ * 	.flatMap(it ->
+ * 		Mono.just(it) // (-)
+ * 	) // (-)
+ * 	.flatMapMany(it ->
+ * 		asyncPublisher(it) // (+)(*)
+ * 	) // (-)
+ * 	.filter(it -> true) // (-)
+ * 	.publishOn(Schedulers.single()) //(+)
+ * 	.map(it -> it) // (-)
+ * 	.map(it -> it * 10) // (-)
+ * 	.filter(it -> true) // (-)
+ * 	.scan((l, r) -> l + r) // (-)
+ * 	.doOnNext(it -> { // (-)
+ * 		//log
+ * 	})
+ * 	.doFirst(() -> { // (-)
+ * 		//log
+ * 	})
+ * 	.doFinally(signalType -> { // (-)
+ * 		//log
+ * 	})
+ * 	.subscribeOn(Schedulers.parallel()) //(+)
+ * 	.subscribe();//(*)
+ * 	(*) - captures brave context at subscription and propagates it
+ * 	(+) - is ASYNC need to decorate Publisher
+ * 	(-) - is SYNC no need to decorate Publisher
+ *}</pre> So it creates 13 (out of 16) less Publisher (+ Subscriber and all their stuff)
+ * on assembly time comparing to the original logic (decorate every Publisher and checking
+ * only at subscription time).
+ * <p/>
+ * It also handles the case when onEachHook was not applied for some operator in the chain
+ * (It could be possible if custom operators or Processor are used) by checkin source
+ * source Publisher. <pre>{@code
+ * 	processor/customOperator // (?) is async and hook is not applied
+ * 	.map(it -> ...) // (+) is SYNC but should add hook as previous Processor/operator does not use hooks
+ * 	.doOnNext(it -> { // (-) is SYNC no need to wrap
+ * 		//log
+ * 	})
+ * 	.subscribe();
+ *}</pre>
+ */
+final class Hack {
+
+	static final Class sourceProducerClass;
+	static {
+		Class c;
+		try {
+			c = Class.forName("reactor.core.publisher.SourceProducer");
+		}
+		catch (ClassNotFoundException e) {
+			c = Void.class;
+		}
+		sourceProducerClass = c;
+	}
+
+	private Hack() {
+	}
+
+	static boolean shouldDecorate(Publisher<?> p) {
+		int maxSteps = 100;
```

Comment:
Please add a comment on this? What's the impact? What would make 10 or 1000 a bad number? are we trying to avoid recursion? or...

Reply:
The point of this is to protect from:  
- reduce overhead of traversal in long chains of `Publisher`s, but it seems would not work most of the time as there are no match reactive chains with such amount of operators. Probably such kind of check would helpful only on `Subscriber` level as their chains are much longer;  
- bad impls which could refers to wrong `source` so it could loop indefinitely;  

100 is just a magic number. The effect was not tested on long chains   
As I mentioned, this is WIP, it stayed here from previous implementation iteration. Probably should be deleted.

Reply:
thanks for the notes.. if this persists, exactly what you wrote would be a fantastic comment!